lib/apis/admin.js

```diff
@@ -337,6 +338,125 @@ module.exports = class AdminAPI {
+
+    // html requests
+    if (req.accepts('html')) {
+      return res.render('admin-dashboard-reports')
+    }
+
+    // fetch
+    var reports = await this.reportsDB.list({
+      cursor: req.query.cursor,
+      limit: req.query.limit ? +req.query.limit : 25,
+      sort: req.query.sort,
+      reverse: +req.query.reverse === 1
+    })
+
+    // respond
+    res.status(200)
+    res.json({reports})
+  }
+
+  async getReport (req, res) {
+    // check perms
+    if (!res.locals.session) throw new UnauthorizedError()
+    if (!res.locals.session.scopes.includes('admin:dats')) throw new ForbiddenError()
+
+    // fetch
+    var report = await this._getReport(req.params.id)
+
+    // respond
+    if (req.accepts('html')) {
+      return res.render('admin-dashboard-report', {record: report})
+    }
+    res.status(200)
+    res.json(report)
+  }
+
+  async updateReport (req, res) {
+    // check perms
+    if (!res.locals.session) throw new UnauthorizedError()
+    if (!res.locals.session.scopes.includes('admin:dats')) throw new ForbiddenError()
+
+    // validate & sanitize input TODO
+    // validate fields that can be changed: status, notes
+    // ;(await req.getValidationResult()).throw()
+
+    // keep a list of changes processed
+    var changedFields = []
+
+    var release = await lock('reports')
+    try {
+      // fetch
+      var report = await this._getReport(req.params.id)
+
+      // update
+      ADMIN_MODIFIABLE_FIELDS_REPORT.forEach(key => {
+        if (typeof req.body[key] !== 'undefined') {
+          changedFields.push(key)
+          report[key] = req.body[key]
+        }
+      })
+
+      await this.reportsDB.put(report)
+    } finally {
+      release()
+    }
+
+    // respond
+    res.status(200)
+    res.json({report, changedFields})
+  }
+
+  async closeReport (req, res) {
+    // check perms
+    if (!res.locals.session) throw new UnauthorizedError()
+    if (!res.locals.session.scopes.includes('admin:dats')) throw new ForbiddenError()
+
+    var release = await lock('reports')
+    try {
+      // fetch
+      var report = await this._getReport(req.params.id)
+
+      // update
+      report.status = 'closed'
+      await this.reportsDB.put(report)
+    } finally {
+      release()
+    }
+    res.status(200)
+    res.json({report})
+  }
+
+  async openReport (req, res) {
+    // check perms
+    if (!res.locals.session) throw new UnauthorizedError()
+    if (!res.locals.session.scopes.includes('admin:dats')) throw new ForbiddenError()
+
+    var release = await lock('reports')
+    try {
+      // fetch
+      var report = await this._getReport(req.params.id)
+
+      // update
+      report.status = 'open'
+      await this.reportsDB.put(report)
+    } finally {
+      release()
+    }
+    res.status(200)
+    res.json({report})
+  }
+
+  async _getReport (id) {
+    return await this.reportsDB.getByID(id)
+    if (!report) throw new NotFoundError()
```

Comment:
This doesnt work the way you're expecting. The return statement will exit the function early, and the `if (!report)` line will never get called